substrate/frame/revive/src/lib.rs

```diff
@@ -547,13 +614,129 @@ pub mod pallet {
+	impl<T: Config> Hooks<BlockNumberFor<T>> for Pallet<T>
+	where
+		// We need this to place the substrate block
+		// hash into the logs of the receipts.
+		T::Hash: frame_support::traits::IsType<H256>,
+		// We need these to access the ETH block gas limit via `Self::evm_block_gas_limit()`.
+		<T as frame_system::Config>::RuntimeCall:
+			Dispatchable<Info = frame_support::dispatch::DispatchInfo>,
+		BalanceOf<T>: Into<U256> + TryFrom<U256>,
+		MomentOf<T>: Into<U256>,
+	{
 		fn on_idle(_block: BlockNumberFor<T>, limit: Weight) -> Weight {
 			let mut meter = WeightMeter::with_limit(limit);
 			ContractInfo::<T>::process_deletion_queue_batch(&mut meter);
 			meter.consumed()
 		}
 
+		fn on_initialize(_n: BlockNumberFor<T>) -> Weight {
+			// TODO: Account for future transactions here in the on_finalize.
+
+			// The pallet executes the following memory operations:
+			//
+			// For every emitted event of type `Event::ContractEmitted`:
+			//  (I) During collection:
+			// 		- read: InflightEvents::<T>::count()
+			// 		- write: InflightEvents::<T>::insert
+			//
+			//	(II) After transaction is executed:
+			// 		- read: InflightEvents::<T>::count() internally by InflightEvents::<T>::drain()
+			// 		- read: Provide the key, value internally by InflightEvents::<T>::drain()
+			// 		- write: Remove key internally by InflightEvents::<T>::drain()
+			// Therefore, we have 2 reads and 2 writes per event, plus one extra read per total
+			// (count).
+			//
+			// Cost(Events) = (2r + 2w) * N + 1r
+			//
+			// For every transaction:
+			//  (I) Cost incurred by the emitted event
+			//
+			//  (II) After transaction is executed:
+			// 		- read: frame_system::Pallet::<T>::extrinsic_index()
+			//  	- read: InflightTransactions::<T>::count()
+			//  	- write: InflightTransactions::<T>::insert()
+			//
+			// Cost(Txs) = Cost(Events) + (1r + 1r) * M + 1r
+			//
+			// On finalize operations:
+			// (I) Operating with transactions:
+			// 		- read: InflightTransactions::<T>::count() internally by
+			//     InflightTransactions::<T>::drain()
+			// 		- read: Provide key, value internally by InflightTransactions::<T>::drain()
+			// 		- write: Remove key internally by InflightTransactions::<T>::drain()
+			//
+			// (II) Parent hash
+			// 		- read: BlockHash::<T>::get()
+			//
+			// (III) Storage propagation
+			// 		- write: BlockHash::<T>::insert() (insert number to hash mapping)
+			// 		- write EthereumBlock::<T>::put() (insert block into storage)
+			// 		- write ReceiptInfoData::<T>::put() (insert receipt data into storage)
+			//
+			// Cost(on_finalize) = Cost(Txs) + (1r + 1r) * M + 1r + 1r + 3w
+			//
+			// Total cost:
+			// (2r + 2w) * N + 1r + (1r + 1r) * M + 1r + (1r + 1r) * M + 1r + 1r + 3w
+			// = 4 M (r + w) + 2 N * (r + w) + 4r + 3w
+			//
+			// Note: This does not account for the cost of computing the state tries.
+
+			Weight::zero()
+		}
+
+		fn on_finalize(block_number: BlockNumberFor<T>) {
```

Comment:
can you add some tests that make sure that all the new storage that we have added in this PR are set / reset as expected when these different hooks are executed

Reply:
I've added a few tests in `revive/src/tests/block_hash` 🙏